### PR TITLE
Cloudstorage WebDAV: fix possible double slash in file URL

### PR DESCRIPTION
**Pull request by SteadyQuad**
Not sure if this is the best place/way for a fix, but it solves my problem. See issue #8805

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8806)

### CODE REVIEW

File: frontend/apps/cloudstorage/webdav.lua [hunk 1]
```diff
@@ -17,7 +17,14 @@ function WebDav:run(address, user, pass, path)
 end
 
 function WebDav:downloadFile(item, address, username, password, local_path, callback_close)
-    local code_response = WebDavApi:downloadFile(address .. WebDavApi:urlEncode( item.url ), username, password, local_path)
+    local urlencoded_item_url = WebDavApi:urlEncode( item.url )
+    local download_address = address
+    if string.sub(address, -1 ) == "/" and string.sub(urlencoded_item_url,1,1) == "/" then
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/webdav.lua:22]: There's no a protocol at this point?

I'll take a look later this week, ping me if I forget. Perhaps this plugin still does some outdated stuff with luasocket

Comment (Frenzie) [reply frontend/apps/cloudstorage/webdav.lua:22]: Or since you know your way around a bit, see if #6307 applies here.

**Comment by hius07** [comment]
It is interesting to find out where does the leading slash come from.

**Comment by hius07** [comment]
I suspect that for the root folder we get empty `path` in line 63
https://github.com/koreader/koreader/blob/4ccdbc90a573579d626af47908440d3914960427/frontend/apps/cloudstorage/webdavapi.lua#L62-L63
and then we get leading slash in
https://github.com/koreader/koreader/blob/4ccdbc90a573579d626af47908440d3914960427/frontend/apps/cloudstorage/webdavapi.lua#L125

**Comment by hius07** [comment]
I would try to fix it in line 125 with something like
```lua
local item_path = path == "" and item_name or path .. "/" .. item_name
```

**Comment by SteadyQuad** [comment]
> I would try to fix it in line 125 with something like
> 
> ```lua
> local item_path = path == "" and item_name or path .. "/" .. item_name
> ```

Thanks! (Not a lua programmer, but that looks like kind of a ternary operator thingy, nice and compact.)

I read through the rest of the code in listFolder, but I have thouble making sense of the logic.

Several things in this codeblock seem strange to me:
https://github.com/koreader/koreader/blob/4ccdbc90a573579d626af47908440d3914960427/frontend/apps/cloudstorage/webdavapi.lua#L54-L77

Like this one: append a slash to the complete url (address+path) if the address part doesn't have a trailing slash?
https://github.com/koreader/koreader/blob/4ccdbc90a573579d626af47908440d3914960427/frontend/apps/cloudstorage/webdavapi.lua#L75-L77

With a call like this 
`listFolder("https://bla.com/test.php/webdav","u","p","myfolder")`
would give this webdav_url (with trailing slash, logical?):
`https://bla.com/test.php/webdav/myfolder/`
while this:
`listFolder("https://bla.com/test.php/webdav/","u","p","myfolder")`
would yield
`https://bla.com/test.php/webdav/myfolder`
which is correct, but different from the above one (no trailing slash).

I don't get the below lines either: comparing the first 2 characters of a string to a single character. I think it would only match if path == "/" which is probably not what was intended. Then in line 67 path gets assigned itself I think. Not sure what the idea is, maybe strip "//mypath" to "/mypath" or something? 
https://github.com/koreader/koreader/blob/4ccdbc90a573579d626af47908440d3914960427/frontend/apps/cloudstorage/webdavapi.lua#L64-L67

Anyway, I'll look a bit closer to see if I can make more sense of it, this part is probably why KOReader won't list files if I omit the trailing slash in the address fields in the cloudstorage UI, the code seems intended to handle that, but doesn't quite. (maybe a unittest would be a nice addition here)

**Comment by SteadyQuad** [comment]
I fixed (or worked around) a few issues with WebDAV:
- downloading files from WebDAV root folder failed in some cases
- subfolders were not recognized/displayed in some cases
- made handling of slashes in URLs somewhat more robust

The code hasn't become any more elegant with these fixes, but my aim was fixing instead of refactoring.

**Comment by hius07** [comment]
I mean, it seems that changes to webdav.lua can be reverted.
Can you test please.

**Comment by adewalike** [comment]
I am facing this issue #8805 too, hoping this PR be merged soon.

**Comment by hius07** [comment]
@adewalike, can you test the attached fix?
Remove .txt extension and replace koreader/frontend/apps/cloudstorage/webdavapi.lua.
[webdavapi.lua.txt](https://github.com/koreader/koreader/files/8601204/webdavapi.lua.txt)

**Comment by adewalike** [comment]
@hius07 Yeah~ It's working perfectly on my devices. Looking forward to seeing it in the next release.

**Comment by hius07** [comment]
Closed in favour of https://github.com/koreader/koreader/pull/9061.

**Pull request closed without merging**